pkg/objstore/swift/swift.go

```diff
@@ -0,0 +1,314 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package swift implements common object storage abstractions against OpenStack swift APIs.
+package swift
+
+import (
+	"context"
+	"fmt"
+	"github.com/pkg/errors"
+	"io"
+	"math/rand"
+	"os"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/improbable-eng/thanos/pkg/objstore"
+
+	"github.com/go-kit/kit/log"
+	"github.com/gophercloud/gophercloud"
+	"github.com/gophercloud/gophercloud/openstack"
+	"github.com/gophercloud/gophercloud/openstack/objectstorage/v1/containers"
+	"github.com/gophercloud/gophercloud/openstack/objectstorage/v1/objects"
+	"github.com/gophercloud/gophercloud/pagination"
+	"github.com/prometheus/client_golang/prometheus"
+	"gopkg.in/yaml.v2"
+)
+
+const (
+	// Class A operations.
+	opObjectsList  = "objects.list"
+	opObjectInsert = "object.insert"
+
+	// Class B operation.
+	opObjectGet = "object.get"
+
+	// Free operations.
+	opObjectDelete = "object.delete"
+)
+
+// DirDelim is the delimiter used to model a directory structure in an object store bucket.
+const DirDelim = "/"
+
+type swiftConfig struct {
+	AuthUrl       string `yaml:"authUrl"`
+	Username      string `yaml:"username,omitempty"`
+	UserId        string `yaml:"userId,omitempty"`
+	Password      string `yaml:"password"`
+	DomainId      string `yaml:"domainId,omitempty"`
+	DomainName    string `yaml:"domainName,omitempty"`
+	TenantID      string `yaml:"tenantId,omitempty"`
+	TenantName    string `yaml:"tenantName,omitempty"`
+	RegionName    string `yaml:"regionName,omitempty"`
+	ContainerName string `yaml:"containerName"`
+}
+
+type Container struct {
+	logger   log.Logger
+	client   *gophercloud.ServiceClient
+	opsTotal *prometheus.CounterVec
+	name     string
+}
+
+func NewContainer(logger log.Logger, conf []byte, reg prometheus.Registerer) (*Container, error) {
+	var sc swiftConfig
+	if err := yaml.Unmarshal(conf, &sc); err != nil {
+		return nil, err
+	}
+
+	authOpts := gophercloud.AuthOptions{
+		IdentityEndpoint: sc.AuthUrl,
+		Username:         sc.Username,
+		UserID:           sc.UserId,
+		Password:         sc.Password,
+		DomainID:         sc.DomainId,
+		DomainName:       sc.DomainName,
+		TenantID:         sc.TenantID,
+		TenantName:       sc.TenantName,
+
+		// Allow Gophercloud to re-authenticate automatically.
+		AllowReauth: true,
+	}
+
+	provider, err := openstack.AuthenticatedClient(authOpts)
+	if err != nil {
+		return nil, err
+	}
+
+	client, err := openstack.NewObjectStorageV1(provider, gophercloud.EndpointOpts{
+		Region: sc.RegionName,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	c := &Container{
+		logger: logger,
+		client: client,
+		opsTotal: prometheus.NewCounterVec(prometheus.CounterOpts{
+			Name:        "thanos_objstore_swift_container_operations_total",
```

Comment:
Not sure we need this here as the factory then uses `BucketWithMetrics`

Reply:
Means we can also remove the `Inc()` calls

Reply:
removed.